tools/generator/src/Extensions/XcodeScheme+Extensions.swift

```diff
@@ -0,0 +1,167 @@
+// MARK: TargetWithID
+
+extension XcodeScheme {
+    struct TargetWithID {
+        let id: TargetID
+        let target: Target
+    }
+}
+
+extension XcodeScheme.TargetWithID: Equatable {
+    static func == (lhs: XcodeScheme.TargetWithID, rhs: XcodeScheme.TargetWithID) -> Bool {
+        return lhs.id == rhs.id
+    }
+}
+
+extension XcodeScheme.TargetWithID: Comparable {
+    static func < (lhs: XcodeScheme.TargetWithID, rhs: XcodeScheme.TargetWithID) -> Bool {
+        return lhs.target.platform < rhs.target.platform
+    }
+}
+
+// MARK: Resolve TargetIDs
+
+extension XcodeScheme {
+    /// Represents a configuration string (Target.configuration).
+    typealias Configuration = String
+
+    /// Determines the mapping of `LabelValue` to the `TargetID` values based upon the scheme's
+    /// configuration.
+    func resolveTargetIDs(targets: [TargetID: Target]) throws -> [LabelValue: TargetID] {
+        // Get all of the scheme labels
+        let allSchemeLabels = allSchemeLabels
+
+        // Create a list of TargetWithID values for the labels that we care about
+        let allLabelValues = Set(allSchemeLabels.map(\.label))
+        let targetWithIDs = targets
+            // We only need target info for labels explicitly mentioned in the scheme
+            .filter { _, target in allLabelValues.contains(target.label) }
+            .map { id, target in TargetWithID(id: id, target: target) }
+
+        let targetInfoByLabelValue = collectTargetInfoByLabelValue(targetWithIDs: targetWithIDs)
+
+        // Collect top-level labels
+        let topLevelLabelValues = allSchemeLabels.filter(\.isTopLevel).map(\.label)
+
+        // This list of the top-level platforms is sorted with the preferred/best platform first.
+        let topLevelPlatforms = Set(
+            targetWithIDs
+                .filter { topLevelLabelValues.contains($0.target.label) }
+                .map(\.target.platform)
+        ).sorted()
+
+        // For each schemeLabel,
+        var resolvedTargetIDs = [LabelValue: TargetID]()
+        for schemeLabel in allSchemeLabels {
+            guard let targetInfo = targetInfoByLabelValue[schemeLabel.label] else {
+                throw PreconditionError(message: """
+Did not find `targetInfo` for label "\(schemeLabel.label)"
+""")
+            }
+
+            let targetID: TargetID
+            if schemeLabel.isTopLevel {
+                // If schemeLabel is top-level, then get the Target-TargetID with the "best"
+                // platform.
+                targetID = try targetInfo.best().id
+            } else {
+                // If schemeLabel is not top-level, then collect all of the Target-TargetID for the
+                // top-level configurations and select the first one.
+                let targetIDs = topLevelPlatforms.compactMap { targetInfo.byPlatform[$0]?.id }
```

Comment:
This can select the wrong targets if  there are multiple that differ by the same platform, but different minOS or bazel configuration. I think you need to look at all of the deps of the top level targets and find the first dep that matches the label.

Reply:
> I presume that we will use a breadth-first search as to select the `TargetID` that is "closest" to top-level target. Does that sound right?

Yes. And we need to collect these for each top-level target. _Then_ we can select the "best" from _those_.

> Is that an implementation detail or can we rely on it? 

Implementation detail. One day I believe I'll start hashing those values to reduce `spec.json` size. So doing a O(1) lookup of a `Target` from the `TargetID` to get it's `.label` would be the way to go.

Reply:
Because of the way `dependencies` work, it might have unrelated platform versions. An example is an iOS app having a dependency on a watch app. I think the search shouldn't match or dig into dependencies that don't have a matching `Platform`.

Reply:
I have updated the resolution logic to reduce the build tree targets to a set that only includes a top-level platform.